front/lib/models.ts

```diff
@@ -1241,6 +1241,183 @@ DataSource.hasMany(DocumentTrackerChangeSuggestion, {
+      primaryKey: true,
+    },
+    textContent: {
+      type: DataTypes.TEXT,
+      allowNull: false,
+    },
+  },
+  {
+    modelName: "user_message",
+    sequelize: front_sequelize,
+  }
+);
+
+export class AssistantMessage extends Model<
+  InferAttributes<AssistantMessage>,
+  InferCreationAttributes<AssistantMessage>
+> {
+  declare id: number;
+  declare textContent: string | null;
+}
+
+AssistantMessage.init(
+  {
+    id: {
+      type: DataTypes.INTEGER,
+      autoIncrement: true,
+      primaryKey: true,
+    },
+    textContent: {
+      type: DataTypes.TEXT,
+      allowNull: true,
+    },
+  },
+  {
+    modelName: "assistant_message",
+    sequelize: front_sequelize,
+  }
+);
+
+export class Message extends Model<
+  InferAttributes<Message>,
+  InferCreationAttributes<Message>
+> {
+  declare id: number;
+
+  declare conversation_id: number;
+  declare version: number;
+  declare rank: number;
+
+  declare is_deleted: boolean;
+
+  declare parentId: ForeignKey<Message["id"]> | null;
+  declare userMessageId: ForeignKey<UserMessage["id"]> | null;
+  declare assistantMessageId: ForeignKey<AssistantMessage["id"]> | null;
+}
+
+Message.init(
+  {
+    id: {
+      type: DataTypes.INTEGER,
+      autoIncrement: true,
+      primaryKey: true,
+    },
+    version: {
+      type: DataTypes.INTEGER,
+      allowNull: false,
+      defaultValue: 0,
+    },
+    is_deleted: {
+      type: DataTypes.BOOLEAN,
+      allowNull: false,
+      defaultValue: false,
+    },
+    conversation_id: {
+      type: DataTypes.INTEGER,
+      allowNull: false,
+    },
+    rank: {
+      type: DataTypes.INTEGER,
+      allowNull: false,
+    },
+  },
+  {
+    modelName: "message",
+    sequelize: front_sequelize,
+    indexes: [
+      {
+        unique: true,
+        fields: ["version", "conversationId", "rank"],
+      },
+    ],
+    hooks: {
+      // TODO @fontanierh: check if we want to add a Check Constraint (from db.ts ?)
+      beforeValidate: (message) => {
+        if (
+          (message.userMessageId === null) ===
+          (message.assistantMessageId === null)
+        ) {
+          throw new Error(
+            "Exactly one of userMessageId, assistantMessageId must be non-null"
+          );
+        }
+      },
+    },
+  }
+);
+
+Message.belongsTo(Conversation, {
+  foreignKey: "conversationId",
+  onDelete: "CASCADE",
+});
+Conversation.hasMany(Message, { foreignKey: "conversationId" });
+UserMessage.hasOne(Message, {
+  foreignKey: "userMessageId",
+  sourceKey: "id",
+  onDelete: "RESTRICT",
+});
+Message.belongsTo(UserMessage, { foreignKey: "userMessageId" });
+AssistantMessage.hasOne(Message, {
+  foreignKey: "assistantMessageId",
+  sourceKey: "id",
+  onDelete: "RESTRICT",
+});
+Message.belongsTo(AssistantMessage, { foreignKey: "assistantMessageId" });
+Message.belongsTo(Message, { foreignKey: "parent_id", as: "parent" });
+Message.hasMany(Message, { foreignKey: "parent_id", as: "children" });
```

Comment:
@spolu
This foreign key isn't unique so making the assumption that multiple messages can have the same parent. WDYT ?

Reply:
Yes @A @B => multiple message will have the same parent indeed